// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package scope;

service Main {
  rpc getCellColorByFeatures (CellColorByFeaturesRequest) returns (CellColorByFeaturesReply) {}
  rpc getCellAUCValuesByFeatures (CellAUCValuesByFeaturesRequest) returns (CellAUCValuesByFeaturesReply) {}
  rpc getCellMetaData (CellMetaDataRequest) returns (CellMetaDataReply) {}
  rpc getFeatures (FeatureRequest) returns (FeatureReply) {}
  rpc getCoordinates (CoordinatesRequest) returns (CoordinatesReply) {}
  rpc getRegulonMetaData (RegulonMetaDataRequest) returns (RegulonMetaDataReply) {}
  rpc getMarkerGenes (MarkerGenesRequest) returns (MarkerGenesReply) {}
  rpc getMyLooms (MyLoomsRequest) returns (MyLoomsReply) {}
  rpc translateLassoSelection (TranslateLassoSelectionRequest) returns (TranslateLassoSelectionReply) {}
  rpc getCellIDs (CellIDsRequest) returns (CellIDsReply) {}
  rpc doGeneSetEnrichment (GeneSetEnrichmentRequest) returns (stream GeneSetEnrichmentReply) {}
  rpc getVmax (VmaxRequest) returns (VmaxReply) {}
  rpc getUUID (UUIDRequest) returns (UUIDReply) {}
  rpc getRemainingUUIDTime (RemainingUUIDTimeRequest) returns (RemainingUUIDTimeReply) {}
  rpc loomUploaded (LoomUploadedRequest) returns (LoomUploadedReply) {}
  rpc getMyGeneSets (MyGeneSetsRequest) returns (MyGeneSetsReply) {}
  rpc deleteUserFile (DeleteUserFileRequest) returns (DeleteUserFileReply) {}
  rpc downloadSubLoom (DownloadSubLoomRequest) returns (stream DownloadSubLoomReply) {}
  rpc setAnnotationName (SetAnnotationNameRequest) returns (SetAnnotationNameReply) {}
  rpc setLoomHierarchy (SetLoomHierarchyRequest) returns (SetLoomHierarchyReply) {}
  rpc getORCID (getORCIDRequest) returns (getORCIDReply) {}
  rpc getORCIDStatus (getORCIDStatusRequest) returns (getORCIDStatusReply) {}
  rpc setColabAnnotationData (setColabAnnotationDataRequest) returns (setColabAnnotationDataReply) {}
  rpc voteAnnotation (voteAnnotationRequest) returns (voteAnnotationReply) {}
  rpc getNextCluster (getNextClusterRequest) returns (FeatureReply) {} 
  rpc addNewClustering (AddNewClusteringRequest) returns (AddNewClusteringReply) {}
  rpc getClusterOverlaps (GetClusterOverlapsRequest) returns (ClusterOverlaps) {}
  rpc getFeatureLabels (FeatureLabelRequest) returns (FeatureLabelReply) {}
}

message ErrorReply {
  string type=1;
  string message=2;
}

message CellColorByFeaturesRequest {
  string loomFilePath=1;
  repeated string feature=2;
  repeated string featureType=3;
  bool hasLogTransform=4;
  bool hasCpmTransform=5;
  repeated float threshold=6;
  bool scaleThresholded=7;
  repeated Annotation annotation=8;
  repeated float vmax=9;
  repeated float vmin=10;
  string logic=11;
}

message ColorLegend {
  repeated string values=1;
  repeated string colors=2;
}

message CellColorByFeaturesReply {
  bool hasAddCompressionLayer=1;
  bytes compressedColor=2;
  repeated string color=3;
  repeated float vmax=4;
  repeated float maxVmax=5;
  repeated int32 cellIndices=6;
  ColorLegend legend=7;
  ErrorReply error=8;
}

message CellAUCValuesByFeaturesRequest {
  string loomFilePath=1;
  repeated string feature=2;
  repeated string featureType=3;
}

message CellAUCValuesByFeaturesReply{
  repeated float value=1;
}

message FeatureRequest {
  string loomFilePath=1;
  string query=2;
}

message CellMetaDataRequest {
  string loomFilePath=1;
  repeated int32 cellIndices=2; // List of cell indices to return, empty list = all cells
  repeated string selectedGenes=3; // List of genes to return epxression values for
  bool hasLogTransform=4;
  bool hasCpmTransform=5;
  repeated string selectedRegulons=6; // As above, for regulons and AUC values
  repeated int32 clusterings=7; // IDs of clustering values to return per cell
  repeated string annotations=8; // String name of annotations to return (from global metadata)

}

message FeatureReply {
  repeated string feature=1;
  repeated string featureType=2;
  repeated string featureDescription=3;
}

message CoordinatesRequest {
  string loomFilePath=1;
  int32 coordinatesID=2;
  repeated Annotation annotation=3;
  string logic=4;
}


message CoordinatesReply {
  repeated float x=1;
  repeated float y=2;
  repeated int32 cellIndices=3;
}

message Annotation {
  string name=1;
  repeated string values=2;
}

message Coordinate {
  float x=1;
  float y=2;
}

message Edge {
  string source=1;
  string target=2;
}

message Trajectory {
  repeated string nodes=1;
  repeated Edge edges=2;
  repeated Coordinate coordinates=3;
}

message Embedding {
  int32 id=1;
  string name=2;
  Trajectory trajectory=3;
}

message ClusterMarkerMetric {
  string accessor=1;
  string name=2;
  string description=3;
}

message CollabAnnoData {
  string curator_name=1;
  string curator_id=2; 
  int64 timestamp=3;
  string obo_id=4;
  string ols_iri=5;
  string annotation_label=6;
  repeated string markers=7;
  string publication=8;
  string comment=9;
}

message CollabAnnoVoter {
  string voter_name=1;
  string voter_id=2;
  bool voter_hash=3;
}

message CollabAnnoVotes {
  int32 total=1;
  repeated CollabAnnoVoter voters=2;
}

message CellTypeAnnotation {
  CollabAnnoData data=1;
  bool validate_hash=2;
  CollabAnnoVotes votes_for=3;
  CollabAnnoVotes votes_against=4;
}

message ClusterAnnotation {
  int32 id=1;
  string description=2;
  repeated CellTypeAnnotation cell_type_annotation=3;
}

message Clustering {
  int32 id=1;
  string group=2;
  string name=3;
  repeated ClusterMarkerMetric clusterMarkerMetrics=4;
  repeated ClusterAnnotation clusters=5;
}

message CellMetaData {
  repeated Annotation annotations=1;
  repeated Embedding embeddings=2;
  repeated Clustering clusterings=3;
}

message AUCThreshold {
  string name=1;
  float threshold=2;
}

message RegulonGenesMetric {
  string accessor=1;
  string name=2;
  string description=3;
  repeated float values=4;
}

message Regulon {
  repeated string genes=1;
  repeated AUCThreshold autoThresholds=2;
  string defaultThreshold=3;
  string motifName=4;
  repeated RegulonGenesMetric metrics=5;
}

message FileMetaData {
  bool hasRegulonsAUC=1;
  bool hasGeneSets=2;
  bool hasClusterings=3;
  bool hasExtraEmbeddings=4;
  bool hasGlobalMeta=5;
  string species=6;
}

message FeatureValues {
  repeated float features=1;
}

message CellAnnotations {
  repeated string annotations=1;
}

message CellClusters {
  repeated int32 clusters=1;
}

message CellMetaDataReply {
  repeated CellClusters clusterIDs=1;
  repeated FeatureValues geneExpression=2;
  repeated FeatureValues aucValues=3;
  repeated CellAnnotations annotations=4;
}

message RegulonMetaDataRequest {
  string loomFilePath=1;
  string regulon=2;
}

message RegulonMetaDataReply {
  Regulon regulonMeta=1;
}

message MarkerGenesRequest {
  string loomFilePath=1;
  int32 clusteringID=2;
  int32 clusterID=3;
}

message MarkerGenesMetric {
  string accessor=1;
  string name=2;
  string description=3;
  repeated float values=4;
}

message MarkerGenesReply {
  repeated string genes=1;
  repeated MarkerGenesMetric metrics=2;
}

message MyLoomsRequest {
  string UUID=1;
  string loomFile=2;
}

message LoomHeierarchy {
  string L1=1;
  string L2=2;
  string L3=3;

}

message MyLoom {
  string loomFilePath=1;
  string loomDisplayName=2;
  int64 loomSize=3;
  CellMetaData cellMetaData=4;
  FileMetaData fileMetaData=5;
  LoomHeierarchy loomHeierarchy=6;

}

message MyLoomsReply {
  repeated MyLoom myLooms=1;
  bool update=2;
}

message TranslateLassoSelectionRequest {
  string srcLoomFilePath=1;
  string destLoomFilePath=2;
  repeated int32 cellIndices=3;
}

message TranslateLassoSelectionReply {
  repeated int32 cellIndices=1;
}

message CellIDsRequest {
  string loomFilePath=1;
  repeated int32 cellIndices=2;
}

message CellIDsReply {
  repeated string cellIds=1;
}

message GeneSetEnrichmentRequest {
  string loomFilePath=1;
  string geneSetFilePath=2;
  string method=3;
}

message Progress {
  float value=1;
  string status=2;
}

message GeneSetEnrichmentReply {
  Progress progress=1;
  bool isDone=2;
  CellColorByFeaturesReply cellValues=3;
}

message VmaxRequest {
  repeated string loomFilePath=1;
  repeated string feature=2;
  repeated string featureType=3;
  bool hasLogTransform=4;
  bool hasCpmTransform=5;
}

message VmaxReply {
  repeated float vmax=1;
  repeated float maxVmax=2;
}

message UUIDRequest {
  string ip=1;
}

message UUIDReply {
  string UUID=1;
}

message RemainingUUIDTimeRequest {
  string ip=1;
  string UUID=2;
  int64 mouseEvents=3;
}

message RemainingUUIDTimeReply {
  string UUID=1;
  int64 timeRemaining=2;
  bool sessionsLimitReached=3;
  string sessionMode=4;
}

message LoomUploadedRequest {
  string UUID=1;
  string filename=2;
}

message LoomUploadedReply {}

message MyGeneSet {
  string geneSetFilePath=1;
  string geneSetDisplayName=2;
}

message MyGeneSetsRequest {
  string UUID=1;
}

message MyGeneSetsReply {
  repeated MyGeneSet myGeneSets=1;
}

message DeleteUserFileRequest {
  string UUID=1;
  string filePath=2;
  string fileType=3;
}

message DeleteUserFileReply {
  bool deletedSuccessfully=1;
}

message DownloadSubLoomRequest {
  string loomFilePath=1;
  string featureType=2;
  string featureName=3;
  string featureValue=4;
  // enum Operator {
  //   EQ = "==";
  // }
  string operator=5;
  // Operator operator = 4 [default = EQ];
}

message DownloadSubLoomReply {
  string loomFilePath=1;
  int64 loomFileSize=2;
  Progress progress=3;
  bool isDone=4;
  ErrorReply error=5;
}

message SetAnnotationNameRequest {
  string loomFilePath=1;
  int64 clusteringID=2;
  int64 clusterID=3;
  string newAnnoName=4;
} 

message SetAnnotationNameReply {
  bool success=1;
}

message SetLoomHierarchyRequest {
  string loomFilePath=1;
  string newHierarchy_L1=2;
  string newHierarchy_L2=3;
  string newHierarchy_L3=4;
} 

message SetLoomHierarchyReply {
  bool success=1;
}

message getORCIDRequest {
  string auth_code=1;
}

message getORCIDReply {
  string orcid_scope_uuid=1;
  string name=2;
  string orcid_id=3;
  bool success=4;
}

message getORCIDStatusRequest {

}

message getORCIDStatusReply {
  bool active=1;
}

message orcidInfoMessage {
  string orcidName=1;
  string orcidID=2;
  string orcidUUID=3;
}

message setColabAnnotationDataRequest {
  string loomFilePath=1;
  int32 clusteringID=2;
  int32 clusterID=3;
  orcidInfoMessage orcidInfo=4;
  CollabAnnoData annoData=5;
}

message setColabAnnotationDataReply {
  bool success=1;
  string message=2;
}

message voteAnnotationRequest {
  string loomFilePath=1;
  int32 clusteringID=2;
  int32 clusterID=3;
  orcidInfoMessage orcidInfo=4;
  CollabAnnoData annoData=5;
  string direction=6;
}

message voteAnnotationReply {
  bool success=1;
  string message=2;
}

message getNextClusterRequest {
  string loomFilePath=1;
  int32 clusteringID=2;
  int32 clusterID=3;
  string direction=4;
}

message NewClusterInfo {
  repeated string cellIDs=1;
  repeated string clusterIDs=2;
  string clusteringName=3;
}

message AddNewClusteringRequest {
  string loomFilePath=1;
  orcidInfoMessage orcidInfo=2;
  NewClusterInfo clusterInfo=3;
}

message AddNewClusteringReply {
  bool success=1;
  string message=2;
}

message GetClusterOverlapsRequest {
  string loomFilePath=1;
  repeated int32 cellIndices=2;
}

message ClusterOverlaps {
  
  message ClusterOverlap {
    string clustering_name=1;
    string cluster_name=2;
    int32 n_cells=3;
    float cells_in_cluster=4;
    float cluster_in_cells=5;
  }

  repeated ClusterOverlap clusterOverlaps=1;
}

message FeatureLabelRequest {
  string loomFilePath=1;
  sint32 embedding=2;
  string feature=3;
}

message FeatureLabelReply {
  message FeatureLabel {
    string label=1;
    string colour=2;
    Coordinate coordinate=3;
  }

  repeated FeatureLabel labels=1;
}